readthedocs/embed/v3/views.py

```diff
@@ -95,6 +96,10 @@ def _get_page_content_from_storage(self, project, version_slug, filename):
             storage_path,
             relative_filename,
         )
+
+        # Decode encoded URLs (e.g. convert %20 into a whitespace)
+        file_path = urllib.parse.unquote(file_path)
```

Comment:
This step should be done on the filename only, not on the final storage path, otherwise this introduces a path traversal vulnerability.

```python
# ../../private/latest/index.html
filename = "%2E%2E%2F%2E%2E%2Fprivate%2Flatest%2Findex%2Ehtml"

# The join succeeds since the path doesn't have `/` or `..`
storage_path = build_media_storage.join('html', filename)

# We will serve this file at the end, which does path traversal to serve a file the user doesn't have access to.
unquote(storage_path)  #  'html/../../private/latest/index.html'
```

Reply:
I pushed a new commit to `unquote` only the `filename`